Consistency Improvement of Examples (#190)
emschwartz/interledger-rs: examples/simple/README.md

# Guide to Simple Two-Node Payment

## Overview

This example sets up two local Interledger.rs nodes, peers them together, and sends a payment from one to the other. This example does not involve any remote nodes or networks. 

To run the full example, you can use `lit` as described [here](../README.md). Otherwise, you can walk through each step below.

Each of the services write their logs to files found under the `logs` directory. You can run `tail -f logs/node-a.log`, for example, to watch the logs of Node A.

![overview](images/overview.svg)

## Prerequisites

- Make sure you have [Redis](https://redis.io/) installed.
    - Compile and install from the source code
        - [Download the source code here](https://redis.io/download)
    - Install using package managers
        - Ubuntu: run `sudo apt-get install redis-server`
        - macOS: If you use Homebrew, run `brew install redis`

## Instructions

### 1. Build interledger.rs
First of all, let's build interledger.rs. This may take a little while.

```bash

> cargo build --bins

Reviewer: It might be worth figuring out a way to have the md to sh script run commands that are commented out in the markdown file so that we can include lines like `echo "Building interledger.rs... (this may take up to a few minutes)"` that don't show up in the rendered README but make the runnable script a bit nicer

Author: It seems nearly impossible because fenced sections are considered as code blocks but I added a function that normal lines start with `<!--!` and end with `-->` are considered as scripts. `!` has no meaning, just for differentiating from normal comments. Multiple lines are not supported so far.

Author: Also markdown files apparently support another comment format like:
```
[xxx]: # (yyy)
```
but I thought HTML comment format was easier to remember.

Reviewer: It seems like a normal HTML comment wrapping a code block section would work too:
```
<!--
\```bash
echo "hello world"
\```
-->
```

(the backslash is in front of the backticks so that it doesn't match the inner ones with the outer ones and render it incorrectly)

Author: I just modified the script to be compatible with multiple lines like:

```
<!--!
echo script goes here
-->
```

This seems a bit simpler so I think this one is better if you don't mind.

The latest commit realizes almost the same as you originally wrote in `run-all.sh` including stopping services and checking balances twice etc.

You can use the command like: `tail -f logs/redis.log`

### 3. Launch 2 Nodes

```bash
# Turn on debug logging for all of the interledger.rs components and tower-web
export RUST_LOG=tower-web,interledger/.*

# Start both nodes
cargo run --package interledger -- node --config ./node-a/config.yml &> logs/node-a.log &
cargo run --package interledger -- node --config ./node-b/config.yml &> logs/node-b.log &

sleep 1
```

Now the Interledger.rs nodes are up and running!  
You can also watch the logs with: `tail -f logs/node-a.log` or `tail -f logs/node-b.log`.

If you are interested in configuration of both nodes, take a look at the config files.
- [Config of node A](./node-a/config.yml)
- [Config of node B](./node-b/config.yml)

### 4. Configure the Nodes

Let's create accounts on both nodes. The following script sets up accounts for two users, Alice and Bob. It also creates accounts that represent the connection between Nodes A and B.

```bash
# Insert accounts on Node A
# One account represents Alice and the other represents Node B's account with Node A

# Create Alice's account
curl \
    -H "Content-Type: application/json" \
    -H "Authorization: Bearer admin-a" \

>     -d @./node-a/alice.json \

Reviewer: I think we should just include the data here as a string instead of loading it from the file

Author: It becomes more similar to post parameters but hmmm, a bit nicer maybe?

Author: Then should node config `.yml`s be broken down into separate config argument like:
```
$ cargo run --package interledger -- node --config ./node-a/config.yml &> logs/node-a.log &
```
to
```
cargo run --package interledger -- node \n
--ilp_address=example.node-a \n
--secret_seed=xxxxxx \n
&> logs/node-a.log &
```
?

I think thi is common usage and not so bad. Also we can leave `--config` option as is.

Reviewer: Fair point. We could do either one. The node currently supports having the options passed as environment variables so you could do:
```
ILP_ADDRESS=example.node-a \
ILP_SECRET_SEED=xxxx \
...
```

Author: fixed.

Reviewer: Actually I put in [this alias](https://github.com/emschwartz/interledger-rs/blob/7e2ff6304e5fbf240e40a4bcb88225ebe10201da/crates/interledger/src/node.rs#L85) so that it would accept `ILP_ADDRESS` instead of `ILP_ILP_ADDRESS` :wink:

Author: wow... I didn't notice.

> # Guide to Simple Two-Node Payment

Reviewer: ```suggestion
# Simple Two-Node Interledger Payment
> A demo of sending a payment between 2 Interledger.rs nodes without settlement.
```

## Overview

This example sets up two local Interledger.rs nodes, peers them together, and sends a payment from one to the other. This example does not involve any remote nodes or networks. 

> To run the full example, you can use `run-md.sh` as described [here](../README.md). Otherwise, you can walk through each step below.

Reviewer: Link to the script

Each of the services write their logs to files found under the `logs` directory. You can run `tail -f logs/node-a.log`, for example, to watch the logs of Node A.

![overview](images/overview.svg)

## Prerequisites

> - To store data of connectors, we need [Redis](https://redis.io/) installed.

Reviewer: ```suggestion
- To store the connector data, we need [Redis](https://redis.io/) installed.
```

    - Compile and install from the source code
        - [Download the source code here](https://redis.io/download)
    - Install using package managers
        - Ubuntu: run `sudo apt-get install redis-server`
        - macOS: If you use Homebrew, run `brew install redis`

## Instructions

<!--!
printf "Stopping Interledger nodes\n"

if lsof -Pi :6379 -sTCP:LISTEN -t >/dev/null ; then
    redis-cli -p 6379 shutdown
fi

if [ -f dump.rdb ] ; then
    rm -f dump.rdb
fi

if lsof -tPi :7770 ; then
    kill `lsof -tPi :7770`
fi

if lsof -tPi :8770 ; then
    kill `lsof -tPi :8770`
fi

printf "\n"
-->

### 1. Build interledger.rs

> First of all, let's build interledger.rs. This may take a little while.

Reviewer: ```suggestion
First of all, let's build interledger.rs. (This may take a couple of minutes)
```

<!--! printf "Building interledger.rs... (this may take a little while)\n" -->
```bash
cargo build --bins
```

### 2. Launch Redis

```bash
# Create the logs directory if it doesn't already exist
mkdir -p logs

# Start Redis

> redis-server --version > /dev/null || printf "\e[31mUh oh! You need to install redis-server before running this example\e[m\n"

Reviewer: Might be worth putting this in the invisible section. Because this guide tells you to install Redis first, I would expect people to have installed it

Author: Fixed.

redis-server &> logs/redis.log &
```

> To watch what's happening on Redis, or generally when you want to watch logs, use the `tail` command.  

Reviewer: Actually the Redis log doesn't say much of interest

Author: 😅 I'll fix.

Author: Fixed.

You can use the command like: `tail -f logs/redis.log`

### 3. Launch 2 Nodes

```bash
# Turn on debug logging for all of the interledger.rs components and tower-web
export RUST_LOG=tower-web,interledger/.*

> # Start both nodes

Reviewer: ```suggestion
# Start both nodes
# Note that the configuration options can be passed as environment variables
# or saved to a YAML file and passed to the node with the `--config` or `-c` CLI argument
```

ILP_SECRET_SEED=8852500887504328225458511465394229327394647958135038836332350604 \
ILP_ADMIN_AUTH_TOKEN=admin-a \
ILP_REDIS_CONNECTION=redis://127.0.0.1:6379/0 \
ILP_HTTP_ADDRESS=127.0.0.1:7770 \
ILP_BTP_ADDRESS=127.0.0.1:7768 \
ILP_SETTLEMENT_ADDRESS=127.0.0.1:7771 \
ILP_DEFAULT_SPSP_ACCOUNT=0 \
cargo run --package interledger -- node &> logs/node-a.log &

ILP_ILP_ADDRESS=example.node-b \
ILP_SECRET_SEED=1604966725982139900555208458637022875563691455429373719368053354 \
ILP_ADMIN_AUTH_TOKEN=admin-b \
ILP_REDIS_CONNECTION=redis://127.0.0.1:6379/1 \
ILP_HTTP_ADDRESS=127.0.0.1:8770 \
ILP_BTP_ADDRESS=127.0.0.1:8768 \
ILP_SETTLEMENT_ADDRESS=127.0.0.1:8771 \
ILP_DEFAULT_SPSP_ACCOUNT=0 \
cargo run --package interledger -- node &> logs/node-b.log &

sleep 2
```

<!--! printf "The Interledger.rs nodes are up and running!\n\n" -->

Now the Interledger.rs nodes are up and running!  
You can also watch the logs with: `tail -f logs/node-a.log` or `tail -f logs/node-b.log`.

### 4. Configure the Nodes

Let's create accounts on both nodes. The following script sets up accounts for two users, Alice and Bob. It also creates accounts that represent the connection between Nodes A and B.

> 

Reviewer: ```suggestion

See the [HTTP API docs](../../docs/api.md) for the full list of fields that can be set on an account.

```

http://localhost:7770/accounts/0/balance

printf "\nNode B's balance on Node A: "
curl \
-H "Authorization: Bearer admin-a" \
http://localhost:7770/accounts/1/balance

printf "\nNode A's balance on Node B: "
curl \
-H "Authorization: Bearer admin-b" \
http://localhost:8770/accounts/1/balance

printf "\nBob's balance: "
curl \
-H "Authorization: Bearer admin-b" \
http://localhost:8770/accounts/0/balance
```

<!--! printf "\n\n" -->

### 7. Kill All the Services
You can finally kill all the services like:

<!--! printf "Stopping Interledger nodes\n" -->

```bash
if lsof -Pi :6379 -sTCP:LISTEN -t >/dev/null ; then
    redis-cli -p 6379 shutdown
fi

if [ -f dump.rdb ] ; then
    rm -f dump.rdb
fi

if lsof -tPi :7770 ; then
    kill `lsof -tPi :7770`
fi

if lsof -tPi :8770 ; then
    kill `lsof -tPi :8770`
fi
```

<!--! printf "\n" -->

## Troubleshooting

> Uh oh! You need to install redis-server before running this example

You need to install Redis to run `redis-server` command. See [Prerequisites](#prerequisites) section.

> curl: (7) Failed to connect to localhost port 7770: Connection refused

Your interledger.rs node is not running. The reason may be:

> 1. You ran following commands too early. Adjust `sleep 1` command to `sleep 3` or larger.

Reviewer: ```suggestion
1. You tried to insert the accounts before the nodes had time to spin up. Adjust the `sleep 1` command to `sleep 3` or larger.
```

1. You have already running interledger.rs nodes on port `7770`. Stop the nodes and retry.
1. You have some other process running on port `7770`. Stop the process and retry.

To stop the process running on port `7770`, try `` kill `lsof -i:7770 -t` ``. Since this example launches 2 nodes, the port may be other than `7770`. Adjust the port number according to the situation.

## Conclusion

> That's it for this example! You've learned how to set up nodes and connect them, even how to send payments between the nodes.

Reviewer: ```suggestion
That's it for this example! You've learned how to set up Interledger.rs nodes, connect them together, and how to send a payment from one to the other.
```

> This is an easy-to-start one so check out the [other examples](../README.md) for more complex demos that show more of the features of Interledger, including multi-hop routing and cross-currency payments.

Reviewer: ```suggestion
Check out the [other examples](../README.md) for more complex demos that show other features of Interledger, including multi-hop routing and cross-currency payments.
```

- Make sure your Redis is empty
    - You could run `redis-cli flushall` to clear all the data.

## Instructions

<!--!
printf "Stopping Interledger nodes\n"

if lsof -Pi :6379 -sTCP:LISTEN -t >/dev/null ; then
    redis-cli -p 6379 shutdown
fi

if [ -f dump.rdb ] ; then
    rm -f dump.rdb
fi

if lsof -tPi :7770 ; then
    kill `lsof -tPi :7770`
fi

if lsof -tPi :8770 ; then
    kill `lsof -tPi :8770`
fi

printf "\n"
-->

### 1. Build interledger.rs
First of all, let's build interledger.rs. (This may take a couple of minutes)

<!--! printf "Building interledger.rs... (This may take a couple of minutes)\n" -->
```bash
cargo build --bins
```

### 2. Launch Redis

<!--!
redis-server --version > /dev/null || printf "\e[31mUh oh! You need to install redis-server before running this example\e[m\n"
-->

```bash
# Create the logs directory if it doesn't already exist
mkdir -p logs

# Start Redis
redis-server &> logs/redis.log &
```

When you want to watch logs, use the `tail` command. You can use the command like: `tail -f logs/redis.log`

### 3. Launch 2 Nodes

```bash
# Turn on debug logging for all of the interledger.rs components and tower-web

> export RUST_LOG=tower-web,interledger/.*

Reviewer: ```suggestion
export RUST_LOG=interledger/.*
```

# Simple Two-Node Interledger Payment
> A demo of sending a payment between 2 Interledger.rs nodes without settlement.

## Overview

This example sets up two local Interledger.rs nodes, peers them together, and sends a payment from one to the other. This example does not involve any remote nodes or networks. 

To run the full example, you can use [`run-md.sh`](../../scripts/run-md.sh) as described [here](../README.md). Otherwise, you can walk through each step below.

Each of the services write their logs to files found under the `logs` directory. You can run `tail -f logs/node-a.log`, for example, to watch the logs of Node A.

![overview](images/overview.svg)

## Prerequisites

- [The Rust environment](#the-rust-environment)
- [Redis](#redis)

> ### The Rust Environment

Reviewer: I would also change this to just "Rust" -- I'm not sure what you mean by the Rust environment

Author: Because it has many components, I thought it might suitable but I don't stick to it.

fi

if [ -f dump.rdb ] ; then
    rm -f dump.rdb
fi

if lsof -tPi :7770 ; then
    kill `lsof -tPi :7770`
fi

if lsof -tPi :8770 ; then
    kill `lsof -tPi :8770`
fi
```

<!--! printf "\n" -->

## Troubleshooting

> Uh oh! You need to install redis-server before running this example

You need to install Redis to run `redis-server` command. See [Prerequisites](#prerequisites) section.

> curl: (7) Failed to connect to localhost port 7770: Connection refused

Your interledger.rs node is not running. The reason may be:

1. You tried to insert the accounts before the nodes had time to spin up. Adjust the `sleep 1` command to `sleep 3` or larger.
1. You have already running interledger.rs nodes on port `7770`. Stop the nodes and retry.
1. You have some other process running on port `7770`. Stop the process and retry.

To stop the process running on port `7770`, try `` kill `lsof -i:7770 -t` ``. Since this example launches 2 nodes, the port may be other than `7770`. Adjust the port number according to the situation.

## Conclusion

That's it for this example! You've learned how to set up Interledger.rs nodes, connect them together, and how to send a payment from one to the other.

> Check out the [other examples](../README.md) for more complex demos that show other features of Interledger, including multi-hop routing and cross-currency payments.

Reviewer: ```suggestion
Check out the [other examples](../README.md) for more complex demos that show other features of Interledger, including settlement, multi-hop routing, and cross-currency payments.
```
